src/model/storage/gdrive.ts

```diff
@@ -0,0 +1,481 @@
+      const s = document.createElement('script');
+      s.src = src;
+      s.onload = () => {
+        resolve();
+      };
+      s.onerror = reject;
+      document.head.appendChild(s);
+    });
+  }
+
+  private _authenticate() {
+    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+    if (GoogleDriveStorage.DEBUG_SYNC) console.log('SYNC: auth start');
+
+    // Based on https://developers.google.com/identity/oauth2/web/guides/use-token-model
+    // TODO: Do we need to switch to authorization code model so we don't get frequent refresh popups
+    const client = google.accounts.oauth2.initTokenClient({
+      client_id: GoogleDriveStorage.CLIENT_ID,
+      scope: GoogleDriveStorage.API_SCOPE,
+      include_granted_scopes: true,
+      prompt: '',
+      callback: (resp: google.accounts.oauth2.TokenResponse) => {
+        if (!resp.access_token) {
+          console.error('Failed to get access token: ', resp);
+          return;
+        }
+
+        // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+        if (GoogleDriveStorage.DEBUG_SYNC) console.log('SYNC: auth complete');
+
+        void this._browserStorage.then(async (store: Storage) => {
+          this._token = resp.access_token;
+          store.setItem(GoogleDriveStorage.TOKEN_STORAGE_KEY, this._token);
+
+          // Do a first synchronization with this token.
+          this._stateSubject.next(DriveSyncState.NEEDS_SYNC);
+          return this.synchronize();
+        });
+      },
+    });
+    client.requestAccessToken();
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  private _handleRequestFailure(reason: gapi.client.Response<any>) {
+    this._retryCount++;
+
+    if (reason.status === 401 || reason.status === 403) {
+      // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+      if (GoogleDriveStorage.DEBUG_SYNC) console.log('SYNC: Refreshing token');
+      this._token = undefined;
+
+      if (this._retryCount < 3) {
+        this._stateSubject.next(DriveSyncState.DISCONNECTED);
+        this._authenticate();
+      } else {
+        this._stateSubject.next(DriveSyncState.FAILED);
+      }
+    } else {
+      this._stateSubject.next(DriveSyncState.FAILED);
+      console.error('Request failed: ', reason);
+      throw new Error('gDrive request failed with status ' + reason.status);
+    }
+  }
+
+  public async disableSync() {
+    if (!this._token) return;
+    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+    if (GoogleDriveStorage.DEBUG_SYNC) console.log('SYNC: Disabling');
+
+    this._stopBackgroundSync();
+
+    // Get rid of the token immediately - worst case, if revoking it fails, we'll still be disconnected.
+    const oldToken = this._token;
+    this._token = undefined;
+    (await this._browserStorage).removeItem(GoogleDriveStorage.TOKEN_STORAGE_KEY);
+    this._stateSubject.next(DriveSyncState.DISCONNECTED);
```

Comment:
I'm not sure if it matters, but don't you want to do this even before you trash the token?

```suggestion
    this._stopBackgroundSync();
    this._stateSubject.next(DriveSyncState.DISCONNECTED);
    
    // Get rid of the token immediately - worst case, if revoking it fails, we'll still be disconnected.
    const oldToken = this._token;
    this._token = undefined;
    (await this._browserStorage).removeItem(GoogleDriveStorage.TOKEN_STORAGE_KEY);
```

Reply:
P.S. Yes, exactly the possible/desired transition type of question :-/

Reply:
In practice, it probably doesn't matter, but if you're strict about asynchronous operations, then technically, transitioning to DISCONNECTED before the token is gone could mean that another state transition out of DISCONNECTED starts and the token is still kept/used, or even that it gets revoked halfway through another sync (very unlikely given that the await won't ever block in practice).

Reply:
Thanks for the explanation, I only started to get it towards the end of the review... makes sense.